// @progress/kendo-react-dateinputs/utils.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { getDate as f, cloneDate as v } from "@progress/kendo-date-math";
import { EMPTY_SELECTIONRANGE as F } from "./calendar/models/SelectionRange.mjs";
function W(e) {
  const n = (r, o, s, ...u) => o[s] === null ? null : (r ? e.isRequired : e)(o, s, ...u), t = n.bind(null, !1);
  return t.isRequired = n.bind(null, !0), t;
}
const d = (e, n, t) => n === void 0 || t === void 0 || n <= e && e <= t ? e : e < n ? n : t, k = new Date(1980, 0, 1), H = new Date(1900, 0, 1), L = new Date(2099, 11, 31), q = new Date(1980, 0, 1), G = new Date(1980, 0, 1, 23, 59, 59), O = (e, n) => {
  const t = v(e);
  return t.setHours(n.getHours(), n.getMinutes(), n.getSeconds(), n.getMilliseconds()), t;
}, X = () => f(/* @__PURE__ */ new Date()), j = (e, n, t) => !e || !(n && n > e || t && t < e), P = (e, n, t) => e === null || !(n && f(n) > f(e) || t && f(t) < f(e)), Y = (e, n) => {
  const { start: t, end: r } = n || F;
  return !t || !r ? !1 : t < e && e < r;
}, z = (e, n, t = 1) => {
  const r = [];
  for (let o = e; o < n; o = o + t)
    r.push(o);
  return r;
}, B = (e, n, t) => n.getTime() <= e.getTime() && e.getTime() <= t.getTime(), J = (e, n) => e.slice(n).concat(e.slice(0, n)), K = (e, n, t) => e && (n && e < n ? v(n) : t && e > t ? v(t) : e), Q = (e) => (n, t = "", r = {}) => {
  const o = document.createElement(e);
  return o.className = t, Object.keys(r).map((s) => o.style[s] = r[s]), typeof n == "string" ? o.innerHTML = n || "" : (n || []).forEach((s) => s && o.appendChild(s)), o;
};
function U(e, n, t = {}) {
  let r, o;
  t.maxWait;
  let s, u, l;
  const g = window, M = !1, h = !1, D = !n && n !== 0 && typeof g.requestAnimationFrame == "function";
  if (typeof e != "function")
    throw new TypeError("Expected a function");
  n = +n || 0;
  function I(i) {
    const c = r, m = o;
    return r = o = void 0, s = e.apply(m, c), s;
  }
  function T(i, c) {
    return D ? (g.cancelAnimationFrame(u), g.requestAnimationFrame(i)) : setTimeout(i, c);
  }
  function R(i) {
    if (D)
      return g.cancelAnimationFrame(i);
    clearTimeout(i);
  }
  function b(i) {
    return u = T(E, n), M ? I() : s;
  }
  function N(i) {
    const c = i - l;
    return n - c;
  }
  function p(i) {
    const c = i - l;
    return l === void 0 || c >= n || c < 0 || h;
  }
  function E() {
    const i = Date.now();
    if (p(i))
      return A();
    u = T(E, N(i));
  }
  function A(i) {
    return u = void 0, r ? I() : (r = o = void 0, s);
  }
  function w() {
    u !== void 0 && R(u), r = l = o = u = void 0;
  }
  function y() {
    return u === void 0 ? s : A();
  }
  function C() {
    return u !== void 0;
  }
  function a(...i) {
    const c = Date.now(), m = p(c);
    return r = i, o = this, l = c, m && u === void 0 ? b() : (u === void 0 && (u = T(E, n)), s);
  }
  return a.cancel = w, a.flush = y, a.pending = C, a;
}
export {
  L as MAX_DATE,
  G as MAX_TIME,
  k as MIDNIGHT_DATE,
  H as MIN_DATE,
  q as MIN_TIME,
  K as dateInRange,
  U as debounce,
  Q as domContainerFactory,
  X as getToday,
  B as intersects,
  P as isInDateRange,
  j as isInRange,
  Y as isInSelectionRange,
  W as nullable,
  z as range,
  O as setTime,
  J as shiftWeekNames,
  d as viewInRange
};
