// @progress/kendo-react-animation/Expand.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as m from "react";
import t from "prop-types";
import { Animation as u } from "./Animation.mjs";
const x = (r) => {
  const {
    appear: i = n.appear,
    enter: e = n.enter,
    exit: o = n.exit,
    transitionEnterDuration: a = n.transitionEnterDuration,
    transitionExitDuration: s = n.transitionExitDuration,
    direction: p = n.direction,
    children: c,
    ...d
  } = r, l = {
    transitionName: `expand-${p}`
  };
  return /* @__PURE__ */ m.createElement(
    u,
    {
      ...l,
      appear: i,
      enter: e,
      exit: o,
      transitionEnterDuration: a,
      transitionExitDuration: s,
      ...d
    },
    c
  );
}, n = {
  appear: !1,
  enter: !0,
  exit: !0,
  transitionEnterDuration: 300,
  transitionExitDuration: 300,
  direction: "vertical"
};
x.propTypes = {
  children: t.oneOfType([t.arrayOf(t.node), t.node]),
  childFactory: t.any,
  className: t.string,
  direction: t.oneOf(["horizontal", "vertical"]),
  component: t.string,
  id: t.string,
  style: t.any
};
export {
  x as Expand
};
