// @progress/kendo-react-data-tools/navigation/constants.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
const e = "data-keyboardnavlevel", a = "data-keyboardnavscope", t = "data-keyboardnavheader", A = "data-keyboardnavbody", o = "data-keyboardnavid", _ = "data-keyboardnavzone", n = "_filter", d = "k-grid-cancel-command", r = "k-grid-edit-command", c = {
  [a]: !0
}, D = {
  [t]: !0
}, E = {
  [A]: !0
};
export {
  d as KEYBOARD_NAV_CANCEL_BUTTON_CLASS,
  A as KEYBOARD_NAV_DATA_BODY,
  t as KEYBOARD_NAV_DATA_HEADER,
  o as KEYBOARD_NAV_DATA_ID,
  e as KEYBOARD_NAV_DATA_LEVEL,
  a as KEYBOARD_NAV_DATA_SCOPE,
  _ as KEYBOARD_NAV_DATA_ZONE,
  r as KEYBOARD_NAV_EDIT_BUTTON_CLASS,
  n as KEYBOARD_NAV_FILTER_COL_SUFFIX,
  E as tableKeyboardNavigationBodyAttributes,
  D as tableKeyboardNavigationHeaderAttributes,
  c as tableKeyboardNavigationScopeAttributes
};
