// @progress/kendo-react-labels/Hint.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as a from "react";
import t from "prop-types";
import { validatePackage as d, useUnstyled as c, classNames as l, uHint as m } from "@progress/kendo-react-common";
import { packageMetadata as p } from "./package-metadata.mjs";
const o = (n) => {
  const e = {
    direction: "start",
    ...n
  };
  d(p);
  const s = c(), i = s && s.uHint, r = a.useMemo(
    () => l(
      m.wrapper({
        c: i,
        direction: e.direction,
        disabled: e.editorDisabled === !0
      }),
      e.className
    ),
    [i, e.direction, e.editorDisabled, e.className]
  );
  return /* @__PURE__ */ a.createElement("div", { id: e.id, style: e.style, className: r }, e.children);
};
o.propTypes = {
  id: t.string,
  direction: t.oneOf(["start", "end"]),
  children: t.oneOfType([t.element, t.node]),
  style: t.object,
  className: t.string,
  editorDisabled: t.bool
};
o.displayName = "KendoReactHint";
export {
  o as Hint
};
