// @progress/kendo-dateinputs-common/dist/npm/common/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var kendo_date_math_1 = require("@progress/kendo-date-math");
var isObject = function (value) { return value && typeof (value) === "object" && !Array.isArray(value); };
var isHtmlElement = function (element) { return element instanceof HTMLElement; };
var dateSetter = function (method) { return function (date, value) {
    var clone = kendo_date_math_1.cloneDate(date);
    clone[method](value);
    return clone;
}; };
/**
 * @hidden
 */
exports.isPresent = function (value) { return value !== undefined && value !== null; };
/**
 * @hidden
 */
exports.isDocumentAvailable = function () { return !!document; };
/**
 * @hidden
 */
exports.isNumber = function (value) { return exports.isPresent(value) && typeof (value) === "number" && !Number.isNaN(value); };
/**
 * @hidden
 */
exports.parseToInt = function (value) { return parseInt(value, 10); };
/**
 * @hidden
 */
exports.isParseableToInt = function (value) { return exports.isNumber(exports.parseToInt(value)) && /^[0-9]+$/.test(value); };
/**
 * @hidden
 */
exports.clamp = function (value, min, max) { return Math.min(max, Math.max(min, value)); };
/**
 * @hidden
 */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
exports.extend = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return Object.assign.apply(Object, args);
};
/**
 * @hidden
 */
exports.deepExtend = function (target) {
    var _a, _b;
    var sources = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        sources[_i - 1] = arguments[_i];
    }
    if (!sources.length) {
        return target;
    }
    var source = sources.shift();
    if (isObject(target) && isObject(source)) {
        for (var key in source) {
            if (isHtmlElement(source[key])) {
                target[key] = source[key];
            }
            else if (isObject(source[key]) && !(source[key] instanceof Date)) {
                if (!target[key] || !isObject(target[key])) {
                    exports.extend(target, (_a = {}, _a[key] = {}, _a));
                }
                exports.deepExtend(target[key], source[key]);
            }
            else {
                exports.extend(target, (_b = {}, _b[key] = source[key], _b));
            }
        }
    }
    return exports.deepExtend.apply(void 0, [target].concat(sources));
};
/**
 * @hidden
 */
exports.noop = function () { };
/**
 * @hidden
 */
exports.isFunction = function (fn) { return typeof (fn) === "function"; };
/**
 * @hidden
 */
exports.cropTwoDigitYear = function (date) {
    if (!exports.isPresent(date) || isNaN(date.getTime())) {
        return 0;
    }
    return Number(date
        .getFullYear()
        .toString()
        .slice(-2));
};
/**
 * @hidden
 */
exports.setYears = dateSetter('setFullYear');
/**
 * @hidden
 */
exports.millisecondDigitsInFormat = function (format) {
    var result = format && format.match(/S+(\1)/);
    return result ? result[0].length : 0;
};
/**
 * @hidden
 */
exports.millisecondStepFor = function (digits) {
    return Math.pow(10, 3 - digits);
};
/**
 * @hidden
 */
exports.areDatePartsEqualTo = function (date, year, month, day, hour, minutes, seconds, milliseconds) {
    if (date &&
        date.getFullYear() === year &&
        date.getMonth() === month &&
        date.getDate() === day &&
        date.getHours() === hour &&
        date.getMinutes() === minutes &&
        date.getSeconds() === seconds &&
        date.getMilliseconds() === milliseconds) {
        return true;
    }
    return false;
};
/**
 * @hidden
 */
exports.isValidDate = function (value) { return exports.isPresent(value) && value.getTime && exports.isNumber(value.getTime()); };
