// @progress/kendo-react-pdf-viewer/PDFViewer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import l from "prop-types";
import { ButtonGroup as Q, Button as d, Toolbar as et, ToolbarSpacer as tt } from "@progress/kendo-react-buttons";
import { TextBox as ot, InputSeparator as at, InputSuffix as nt } from "@progress/kendo-react-inputs";
import { Pager as rt } from "@progress/kendo-react-data-tools";
import { Upload as lt } from "@progress/kendo-react-upload";
import { ComboBox as ct } from "@progress/kendo-react-dropdowns";
import { Loader as st } from "@progress/kendo-react-indicators";
import { validatePackage as it, shouldShowValidationUI as ut, classNames as mt, WatermarkOverlay as dt } from "@progress/kendo-react-common";
import { useLocalization as gt } from "@progress/kendo-react-intl";
import { zoomOutIcon as ft, zoomInIcon as ht, pointerIcon as vt, handIcon as bt, searchIcon as kt, folderOpenIcon as pt, downloadIcon as yt, printIcon as Ct, convertLowercaseIcon as Et, arrowUpIcon as St, arrowDownIcon as xt, xIcon as wt } from "@progress/kendo-svg-icons";
import "pdfjs-dist/build/pdf.worker.min.mjs";
import { packageMetadata as X } from "./package-metadata.mjs";
import { Scroller as Zt, SearchService as Lt, scrollToPage as Y, removeChildren as z, loadPDF as ee, reloadDocument as zt, goToNextSearchMatch as te, goToPreviousSearchMatch as It, currentPage as Nt, calculateZoomLevel as Tt, download as Mt, print as Pt, DEFAULT_ZOOM_LEVEL as Rt } from "@progress/kendo-pdfviewer-common";
import { messages as i, zoomOut as oe, zoomIn as ae, zoomLevel as ne, enableSelection as re, enablePanning as le, search as I, open as ce, download as se, print as ie, matchCase as ue, searchOf as me, prevMatch as de, nextMatch as ge, close as fe, actualWidth as Ft, fitToWidth as Ot, fitToPage as Dt, popupBlocked as he } from "./messages.mjs";
const be = [
  "pager",
  "spacer",
  "zoomInOut",
  "zoom",
  "selection",
  "spacer",
  "search",
  "open",
  "download",
  "print"
], y = {
  minZoom: 0.5,
  maxZoom: 4,
  tools: [...be],
  zoomRate: 0.25,
  zoomLevels: [
    { id: 1, priority: 1, value: 1, text: "Actual width", type: "ActualWidth", locationString: Ft },
    { id: 2, priority: 2, value: 1, text: "Fit to width", type: "FitToWidth", locationString: Ot },
    { id: 3, priority: 3, value: 1, text: "Fit to page", type: "FitToPage", locationString: Dt },
    { id: 4, priority: 4, value: 0.5, text: "50%", type: "" },
    { id: 5, priority: 5, value: 0.75, text: "75%", type: "" },
    { id: 6, priority: 100, value: 1, text: "100%", type: "" },
    { id: 7, priority: 7, value: 1.25, text: "125%", type: "" },
    { id: 8, priority: 8, value: 1.5, text: "150%", type: "" },
    { id: 9, priority: 9, value: 2, text: "200%", type: "" },
    { id: 10, priority: 10, value: 3, text: "300%", type: "" },
    { id: 11, priority: 11, value: 4, text: "400%", type: "" }
  ],
  defaultZoom: Rt
}, Bt = [
  ".k-toolbar > button",
  ".k-toolbar .k-combobox > input",
  ".k-toolbar .k-button-group > button",
  ".k-toolbar .k-pager > a",
  ".k-toolbar .k-pager input"
];
let ve = 0;
function At() {
  return ve++, ve;
}
const Vt = (a, E) => a.priority > E.priority ? -1 : a.priority < E.priority ? 1 : 0, ke = e.forwardRef((a, E) => {
  it(X);
  const pe = ut(X), {
    zoom: D,
    zoomLevels: B = y.zoomLevels,
    defaultZoom: N = y.defaultZoom,
    minZoom: T = y.minZoom,
    maxZoom: M = y.maxZoom,
    zoomRate: S = y.zoomRate
  } = a, s = gt(), c = e.useRef(null), [ye, x] = e.useState(N), g = D !== void 0 ? D : ye, C = B.slice().sort(Vt).find((o) => o.value === g) || {
    text: g * 100 + "%",
    value: g,
    id: g,
    locationString: ""
  };
  C.locationString && (C.text = s.toLanguageString(
    C.locationString,
    i[C.locationString]
  ));
  const [u, P] = e.useState(!1), [Ce, v] = e.useState(!0), [w, A] = e.useState(0), [Z, V] = e.useState(!0), [Ee, W] = e.useState(!1), [b, R] = e.useState(0), [f, k] = e.useState(0), [p, K] = e.useState(!1), [F, U] = e.useState(""), t = e.useMemo(() => ({}), []);
  t.currentZoom = g, t.props = a;
  const j = e.useCallback((o) => {
    c.current && c.current.style.setProperty("--scale-factor", String(o));
  }, []), h = e.useRef(null), L = e.useRef(null);
  e.useImperativeHandle(
    h,
    () => ({
      get element() {
        return L.current;
      },
      props: a,
      get pages() {
        return t.pages;
      },
      get document() {
        return t.document;
      }
    }),
    []
  ), e.useImperativeHandle(E, () => h.current);
  const Se = e.useCallback(() => {
    if (t.props.onLoad) {
      const o = { target: h.current };
      t.props.onLoad.call(void 0, o);
    }
  }, []), q = e.useCallback(
    (o, n, r) => {
      if (a.onDownload) {
        const m = {
          target: h.current,
          blob: o,
          fileName: n,
          saveOptions: r
        };
        return a.onDownload.call(void 0, m) === !1;
      }
      return !1;
    },
    [a.onDownload]
  ), xe = e.useCallback(() => {
    var o;
    t.scroller && t.scroller.destroy(), t.scroller = new Zt((o = c.current) == null ? void 0 : o.parentNode, {
      filter: ".k-page",
      events: {}
    }), t.scroller.disablePanEventsTracking();
  }, []), we = e.useCallback((o) => {
    const n = Array.from(o.querySelectorAll(".k-text-layer"));
    t.search = new Lt({
      textContainers: n || [],
      highlightClass: "k-search-highlight",
      highlightMarkClass: "k-search-highlight-mark",
      charClass: "k-text-char"
    });
  }, []);
  t.done = e.useCallback(({ pdfPages: o, pdfDoc: n, zoom: r }) => {
    t.document = n, t.pages = o, t.zoom = r, xe(), v(!1), P(!0), Se(), c.current && Y(c.current, 0);
  }, []), t.error = e.useCallback(
    (o) => {
      if (t.document = null, t.pages = [], v(!1), P(!1), a.onError) {
        const n = {
          error: typeof o == "string" ? { message: o } : o,
          target: h.current
        };
        a.onError.call(void 0, n);
      }
    },
    [a.onError]
  ), e.useEffect(() => {
    c.current && (a.url || a.data || a.arrayBuffer ? (v(!0), z(c.current), ee({
      url: a.url,
      data: a.data,
      arrayBuffer: a.arrayBuffer,
      dom: c.current,
      zoom: t.currentZoom,
      done: t.done,
      error: t.error
    }), j(t.currentZoom)) : (t.document = null, t.pages = [], P(!1), v(!1), z(c.current)));
  }, [a.url, a.data, a.arrayBuffer]);
  const H = e.useCallback((o, n) => {
    c.current && (v(!0), z(c.current), zt({
      pdfDoc: o,
      zoom: n,
      dom: c.current,
      done: (r) => {
        t.pages = r, t.zoom = n, v(!1);
      },
      error: t.error
    }));
  }, []);
  e.useEffect(() => {
    j(g), c.current && t.document && g !== t.zoom && H(t.document, g);
  }, [g, H]), e.useEffect(() => () => {
    t.scroller && t.scroller.destroy(), t.search && t.search.destroy(), t.document = null, t.pages = [];
  }, []);
  const Ze = e.useCallback(() => {
    W(!0), we(c.current);
  }, []), Le = e.useCallback(
    (o) => {
      const n = o.value, r = t.search.search({ text: n, matchCase: p });
      k(r.length ? 1 : 0), R(r.length), U(n);
    },
    [p]
  ), ze = e.useCallback(() => {
    const o = t.search.search({ text: F, matchCase: !p });
    k(o.length ? 1 : 0), R(o.length), K(!p);
  }, [p, F]), Ie = e.useCallback(() => {
    te(t), k(f + 1 > b ? 1 : f + 1);
  }, [f, b]), Ne = e.useCallback(() => {
    It(t), k(f - 1 < 1 ? b : f - 1);
  }, [f, b]), $ = e.useCallback(() => {
    t.search.destroy(), k(0), R(0), K(!1), U(""), W(!1);
  }, []), Te = e.useCallback(
    (o) => {
      o.key === "Enter" ? (o.preventDefault(), te(t), k(f + 1 > b ? 1 : f + 1)) : o.key === "Escape" && $();
    },
    [f, b]
  ), Me = e.useCallback(
    (o) => {
      if (c.current) {
        const n = o.skip;
        Y(c.current, n);
        const r = {
          page: n + 1,
          target: h.current,
          syntheticEvent: o.syntheticEvent
        };
        a.onPageChange && a.onPageChange.call(void 0, r);
      }
      A(o.skip);
    },
    [w, a.onPageChange]
  ), Pe = e.useCallback(
    (o) => {
      if (L.current) {
        const n = Nt(L.current);
        if (n !== w) {
          A(n);
          const r = {
            page: n + 1,
            target: h.current,
            syntheticEvent: o
          };
          a.onPageChange && a.onPageChange.call(void 0, r);
        }
      }
    },
    [w, a.onPageChange]
  ), Re = e.useCallback(
    (o) => {
      const n = Math.min(t.currentZoom + S, M);
      if (n !== t.currentZoom && t.document && (x(n), a.onZoom)) {
        const r = {
          zoom: n,
          target: h.current,
          syntheticEvent: o
        };
        a.onZoom.call(void 0, r);
      }
    },
    [S, M, a.onZoom]
  ), Fe = e.useCallback(
    (o) => {
      const n = Math.max(t.currentZoom - S, T);
      if (n !== t.currentZoom && t.document && (x(n), a.onZoom)) {
        const r = {
          zoom: n,
          target: h.current,
          syntheticEvent: o
        };
        a.onZoom.call(void 0, r);
      }
    },
    [S, T, a.onZoom]
  ), Oe = e.useCallback(
    (o) => {
      const n = o.value === null ? { text: "100%", value: 1, id: 100 } : { ...o.value };
      if (n.value === void 0) {
        const m = parseFloat(n.text);
        typeof m == "number" && !Number.isNaN(m) ? n.value = m / 100 : n.value = 1;
      }
      let r = n ? Tt(n.value, n.type, t.currentZoom, c.current) : 1;
      if (r = Math.round(r * 100) / 100, t.currentZoom !== r && t.document && (x(r), a.onZoom)) {
        const m = {
          zoom: r,
          target: h.current,
          syntheticEvent: o.syntheticEvent
        };
        a.onZoom.call(void 0, m);
      }
    },
    [a.onZoom]
  ), De = e.useCallback(() => {
    t.scroller.disablePanEventsTracking(), V(!0);
  }, []), Be = e.useCallback(() => {
    t.scroller.enablePanEventsTracking(), V(!1);
  }, []), Ae = e.useCallback(() => {
    Mt(
      {
        pdf: t.document,
        error: t.error
      },
      a.saveFileName,
      a.saveOptions,
      q
    );
  }, [a.url, a.data, a.arrayBuffer, a.saveFileName, a.saveOptions, q]), Ve = e.useCallback(() => {
    v(!0);
    const o = (r) => {
      t.error(
        typeof (r ? r.message || r : null) == "string" ? r.message || r : s.toLanguageString(he, i[he])
      );
    }, n = () => {
      v(!1);
    };
    Pt(t.pages, n, o);
  }, []), We = e.useCallback(
    (o) => {
      const n = o.newState;
      n[0] && n[0].getRawFile && n[0].getRawFile().arrayBuffer().then((m) => {
        if (c.current) {
          v(!0), z(c.current);
          const O = t.props.zoom === void 0 ? N : t.props.zoom;
          ee({
            arrayBuffer: m,
            dom: c.current,
            zoom: O,
            done: t.done,
            error: t.error
          }), x(O);
        }
      });
    },
    [N]
  ), Ke = e.useCallback((o) => {
    const n = o.target;
    if (n instanceof Element && n.parentNode) {
      const r = n.closest(".k-toolbar"), m = r && r.querySelector(".k-upload input");
      m && m.click();
    }
  }, []), _ = Ce && /* @__PURE__ */ e.createElement("div", { className: "k-loader-container k-loader-container-md k-loader-top" }, /* @__PURE__ */ e.createElement("div", { className: "k-loader-container-overlay k-overlay-light" }), /* @__PURE__ */ e.createElement("div", { className: "k-loader-container-inner " }, /* @__PURE__ */ e.createElement(st, { size: "large" }))), Ue = /* @__PURE__ */ e.createElement(Q, { className: "k-toolbar-button-group k-button-group-solid" }, /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-group-start",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(oe, i[oe]),
      disabled: g <= T || !u,
      onClick: Fe,
      icon: "zoom-out",
      svgIcon: ft
    }
  ), /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-group-end",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(ae, i[ae]),
      disabled: g >= M || !u,
      onClick: Re,
      icon: "zoom-in",
      svgIcon: ht
    }
  )), je = /* @__PURE__ */ e.createElement(
    ct,
    {
      className: "k-toolbar-combobox",
      disabled: !u,
      data: B.map((o) => ({
        ...o,
        text: o.locationString ? s.toLanguageString(o.locationString, i[o.locationString]) : o.text
      })),
      dataItemKey: "id",
      textField: "text",
      value: u ? C : null,
      allowCustom: !0,
      onChange: Oe,
      placeholder: s.toLanguageString(ne, i[ne])
    }
  ), qe = /* @__PURE__ */ e.createElement(
    rt,
    {
      disabled: !u,
      previousNext: !0,
      type: "input",
      skip: w,
      take: 1,
      total: t.pages ? t.pages.length : 0,
      info: !1,
      onPageChange: Me
    }
  ), He = /* @__PURE__ */ e.createElement(tt, null), $e = /* @__PURE__ */ e.createElement(Q, { className: "k-toolbar-button-group k-button-group-solid" }, /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-group-start",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(re, i[re]),
      icon: "pointer",
      svgIcon: vt,
      disabled: !u,
      togglable: !0,
      selected: Z && u,
      onClick: De
    }
  ), /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-group-end",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(le, i[le]),
      icon: "hand",
      svgIcon: bt,
      disabled: !u,
      togglable: !0,
      selected: !Z && u,
      onClick: Be
    }
  )), _e = /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-toolbar-button",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(I, i[I]),
      icon: "search",
      svgIcon: kt,
      disabled: !u,
      onClick: Ze
    }
  ), Ge = /* @__PURE__ */ e.createElement(e.Fragment, null, /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-toolbar-button",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(ce, i[ce]),
      icon: "folder-open",
      svgIcon: pt,
      onClick: Ke
    }
  ), /* @__PURE__ */ e.createElement("div", { style: { display: "none" } }, /* @__PURE__ */ e.createElement(
    lt,
    {
      restrictions: { allowedExtensions: [".pdf"] },
      onAdd: We,
      autoUpload: !1,
      defaultFiles: [],
      multiple: !1,
      accept: ".pdf,.PDF",
      withCredentials: !1
    }
  ))), Je = /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-toolbar-button",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(se, i[se]),
      icon: "download",
      svgIcon: yt,
      disabled: !u,
      onClick: Ae
    }
  ), Qe = /* @__PURE__ */ e.createElement(
    d,
    {
      className: "k-toolbar-button",
      fillMode: "flat",
      themeColor: "base",
      title: s.toLanguageString(ie, i[ie]),
      icon: "print",
      svgIcon: Ct,
      disabled: !u,
      onClick: Ve
    }
  ), Xe = {
    pager: qe,
    spacer: He,
    zoomInOut: Ue,
    zoom: je,
    selection: $e,
    search: _e,
    open: Ge,
    download: Je,
    print: Qe
  }, Ye = (a.tools || y.tools).map((o) => ({
    ...Xe[o],
    key: "toobar-tool-" + o + At()
  })), G = /* @__PURE__ */ e.createElement(et, { buttons: Bt }, ...Ye), J = /* @__PURE__ */ e.createElement(
    "div",
    {
      className: mt("k-canvas k-pdf-viewer-canvas k-pos-relative k-overflow-auto", {
        "k-enable-text-select": Z,
        "k-enable-panning": !Z
      }),
      onScroll: Pe
    },
    Ee && /* @__PURE__ */ e.createElement("div", { className: "k-search-panel k-pos-sticky k-top-center" }, /* @__PURE__ */ e.createElement(
      ot,
      {
        value: F,
        onChange: Le,
        placeholder: s.toLanguageString(I, i[I]),
        autoFocus: !0,
        onKeyDown: Te,
        suffix: () => /* @__PURE__ */ e.createElement(e.Fragment, null, /* @__PURE__ */ e.createElement(at, null), /* @__PURE__ */ e.createElement(nt, null, /* @__PURE__ */ e.createElement(
          d,
          {
            icon: "convert-lowercase",
            svgIcon: Et,
            title: s.toLanguageString(ue, i[ue]),
            fillMode: "flat",
            togglable: !0,
            selected: p,
            onClick: ze
          }
        )))
      }
    ), /* @__PURE__ */ e.createElement("span", { className: "k-search-matches" }, /* @__PURE__ */ e.createElement("span", null, f), " ", s.toLanguageString(me, i[me]), /* @__PURE__ */ e.createElement("span", null, b)), /* @__PURE__ */ e.createElement(
      d,
      {
        title: s.toLanguageString(de, i[de]),
        fillMode: "flat",
        icon: "arrow-up",
        svgIcon: St,
        disabled: b === 0,
        onClick: Ne
      }
    ), /* @__PURE__ */ e.createElement(
      d,
      {
        title: s.toLanguageString(ge, i[ge]),
        fillMode: "flat",
        icon: "arrow-down",
        svgIcon: xt,
        disabled: b === 0,
        onClick: Ie
      }
    ), /* @__PURE__ */ e.createElement(
      d,
      {
        title: s.toLanguageString(fe, i[fe]),
        fillMode: "flat",
        icon: "x",
        svgIcon: wt,
        onClick: $
      }
    )),
    /* @__PURE__ */ e.createElement("div", { ref: c, className: "k-pdf-viewer-pages" })
  );
  return /* @__PURE__ */ e.createElement("div", { className: "k-pdf-viewer", style: a.style, ref: L }, a.onRenderLoader ? a.onRenderLoader.call(void 0, _ || null) : _, a.onRenderToolbar ? a.onRenderToolbar.call(void 0, G) : G, a.onRenderContent ? a.onRenderContent.call(void 0, J) : J, pe && /* @__PURE__ */ e.createElement(dt, null));
});
ke.displayName = "KendoReactPDFViewer";
ke.propTypes = {
  url: l.string,
  data: l.string,
  arrayBuffer: l.any,
  typedArray: l.any,
  style: l.object,
  saveFileName: l.string,
  saveOptions: l.object,
  tools: l.arrayOf(l.oneOf(be).isRequired),
  zoomLevels: l.arrayOf(l.any),
  zoom: l.number,
  defaultZoom: l.number,
  minZoom: l.number,
  maxZoom: l.number,
  zoomRate: l.number,
  onError: l.func,
  onLoad: l.func,
  onDownload: l.func,
  onRenderToolbar: l.func,
  onRenderContent: l.func,
  onRenderLoader: l.func,
  onZoom: l.func
};
export {
  ke as PDFViewer
};
