// @progress/kendo-react-data-tools/header/HeaderThElement.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import { useTableKeyboardNavigation as f } from "../navigation/hooks.mjs";
import { disableNavigatableContainer as b } from "@progress/kendo-react-common";
const p = (r) => {
  const {
    ariaSort: i,
    colSpan: n,
    rowSpan: o,
    columnId: l,
    navigatable: e,
    ariaColumnIndex: c,
    ariaLabel: s,
    ariaSelected: m,
    ariaDescription: h,
    ...d
  } = r, u = f(l, e), a = t.useRef(null);
  return t.useEffect(() => {
    a.current && e && b(a.current);
  }, [e]), /* @__PURE__ */ t.createElement(
    "th",
    {
      ref: a,
      "aria-sort": i,
      "aria-label": s,
      "aria-colindex": c,
      "aria-selected": m,
      colSpan: n,
      rowSpan: o,
      ...u,
      ...d
    },
    r.children
  );
};
p.displayName = "KendoReactHeaderThElement";
export {
  p as HeaderThElement
};
