// @progress/kendo-pdfviewer-common/dist/npm/common/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Component = void 0;
const core_1 = require("./core");
class Component {
    constructor(element, options) {
        this.options = {
            events: {}
        };
        this.element = element;
        this.options = options;
    }
    destroy() {
        this.element = null;
    }
    extendOptions(options) {
        this.options = (0, core_1.deepExtend)(this.options, options);
    }
    setOptions(options) {
        this.options = options;
    }
    trigger(eventName, args = {}) {
        const eventData = {
            defaultPrevented: false,
            preventDefault: () => {
                eventData.defaultPrevented = true;
            }
        };
        if ((0, core_1.isFunction)(this.options.events[eventName])) {
            this.options.events[eventName]((0, core_1.extend)(eventData, args, {
                sender: this
            }));
        }
        return eventData.defaultPrevented;
    }
}
exports.Component = Component;
