// @progress/kendo-react-dateinputs/dateinput/dateInputIntl.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { localeInfo as l } from "@progress/kendo-intl";
class r {
  constructor(t) {
    this.service = t, this.cldr = {}, this.locale = t.locale, this.format = t.format, this.toString = t.toString, this.formatDate = t.formatDate, this.parseDate = t.parseDate, this.parseNumber = t.parseNumber, this.formatNumber = t.formatNumber, this.splitDateFormat = t.splitDateFormat, this.numberSymbols = t.numberSymbols, this.firstDay = t.firstDay, this.dateFieldName = t.dateFieldName, this.dateFormatNames = (o, s) => this.service.dateFormatNames(s), this.weekendRange = (o) => {
    };
    const a = l(this.locale);
    this.cldr[a.name] = a;
  }
}
export {
  r as DateInputIntl
};
